File: leetcode/leetcode98.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

class Solution {
public:
    bool isValidBST(TreeNode* root)
    {
        return isValidBST(root, root->val, root->val, -1, -1);
    }
    bool isValidBST(TreeNode* root, int min, int max, int mostLeft, int mostRight)
    {
        if (root == nullptr)
            return true;
        if ((mostLeft != -1 && root->val <= min) || (mostRight != -1 && root->val >= max)) {
            return false;
        }
        return isValidBST(root->left, min, root->val, mostLeft, 0)
            && isValidBST(root->right, root->val, max, 0, mostRight);
    }
};
int main(void)
{
    Solution s;
    TreeNode* root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);
    cout << s.isValidBST(root) << endl;
    return 0;
}

File: leetcode/10-99/leetcode98.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


from typing import Optional


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def chaewon(
            root: TreeNode, mi: int, ma: int, most_left: bool, most_right: bool
        ):
            if not root:
                return True
            if most_left and root.val <= mi:
                return False
            if most_right and root.val >= ma:
                return False
            return chaewon(
                root=root.left,
                mi=mi,
                ma=root.val,
                most_left=most_left,
                most_right=True,
            ) and chaewon(
                root=root.right,
                mi=root.val,
                ma=ma,
                most_left=True,
                most_right=most_right,
            )

        return chaewon(root=root, mi=0, ma=0, most_left=False, most_right=False)
